service/history/engine/engineimpl/history_engine.go

```diff
@@ -261,25 +220,13 @@ func NewEngineWithShardContext(
 	)
 	openExecutionCheck := invariant.NewConcreteExecutionExists(pRetry, shard.GetDomainCache())
 
-	historyEngImpl.txProcessor = queueProcessorFactory.NewTransferQueueProcessor(
-		shard,
-		historyEngImpl,
-		queueTaskProcessor,
-		executionCache,
-		historyEngImpl.workflowResetter,
-		historyEngImpl.archivalClient,
-		openExecutionCheck,
-		historyEngImpl.wfIDCache,
-	)
-
-	historyEngImpl.timerProcessor = queueProcessorFactory.NewTimerQueueProcessor(
-		shard,
-		historyEngImpl,
-		queueTaskProcessor,
-		executionCache,
-		historyEngImpl.archivalClient,
-		openExecutionCheck,
-	)
+	for _, factory := range queueFactories {
+		historyEngImpl.queueProcessors[factory.Category()] = factory.CreateQueue(
+			shard,
```

Comment:
shard.SetEngine() is called later in this function but the queue factory is calling shard.GetEngine(). Let's make sure the dependencies are clear and future proof.

Reply:
The queue processors are depending on shard context and history engine. History engine is depending on shard context.